llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,393 @@
+  AspectsSetTy Result;
+
+  for (const Type *TT : T->subtypes()) {
+    AspectsSetTy Aspects = getAspectsFromType(TT, Types);
+    Result.insert(Aspects.begin(), Aspects.end());
+  }
+
+  Types[T] = Result;
+  return Result;
+}
+
+/// Returns aspects which might be used in instruction @I.
+/// Function inspects return type and all operand's types.
+/// NB! This function inserts new records in @Types map for new discovered
+/// types. For the best perfomance pass this map in the next invocations.
+AspectsSetTy getAspectsUsedByInstruction(const Instruction &I,
+                                         TypeToAspectsMapTy &Types) {
+  Type *ReturnType = I.getType();
+  AspectsSetTy Result = getAspectsFromType(ReturnType, Types);
+  for (const auto &OperandIt : I.operands()) {
+    Type *T = OperandIt->getType();
+    AspectsSetTy Aspects = getAspectsFromType(T, Types);
+    Result.insert(Aspects.begin(), Aspects.end());
+  }
+
+  return Result;
+}
+
+/// Class for emiting warning messages. Unfortunately, LLVMContext
+/// doesn't contain tools for that.
+class MissedAspectDiagnosticInfo : public DiagnosticInfo {
+  Twine Msg;
+
+public:
+  MissedAspectDiagnosticInfo(Twine DiagMsg,
+                             DiagnosticSeverity Severity = DS_Warning)
+      : DiagnosticInfo(DK_Linker, Severity), Msg(std::move(DiagMsg)) {}
+
+  void print(DiagnosticPrinter &DP) const override { DP << Msg; }
+};
+
+void emitWarning(LLVMContext &C, const StringRef Msg) {
+  C.diagnose(MissedAspectDiagnosticInfo(Msg));
+}
+
+std::string join(const AspectsSetTy &C, char sep) {
+  std::string S;
+  bool FirstOccurence = true;
+  for (int Aspect : C) {
+    if (!FirstOccurence) {
+      S += sep;
+      S += ' ';
+    }
+
+    FirstOccurence = false;
+    S += std::to_string(Aspect);
+  }
+
+  return S;
+}
+
+/// Checks that all declared function's aspects correspond to the
+/// aspects set @Aspects. If there is a inconsistency then corresponding
+/// warning is emitted.
+template <class Container>
+void checkDeclaredAspectsForFunction(LLVMContext &C, const Function *F,
+                                     const Container &UsedAspects) {
+  MDNode *MDN = F->getMetadata("intel_declared_aspects");
+  if (!MDN)
+    return;
+
+  AspectsSetTy DeclaredAspects;
+  for (auto &OperandIt : MDN->operands()) {
+    const auto *CAM = dyn_cast<const ConstantAsMetadata>(OperandIt);
+    assert(CAM &&
+           "constant are expected in intel_declared_aspects list's entries");
+    DeclaredAspects.insert(
+        cast<const ConstantInt>(CAM->getValue())->getSExtValue());
+  }
+
+  AspectsSetTy MissedAspects;
+  for (int Aspect : UsedAspects) {
+    if (DeclaredAspects.count(Aspect) == 0)
+      MissedAspects.insert(Aspect);
+  }
+
+  if (!MissedAspects.empty()) {
+    std::string AspectsStr = join(MissedAspects, ',');
+    // TODO: demangle function name and aspect's IDs?
+    std::string Msg = formatv(
+        "for function \"{0}\" there is the list of missed aspects: [{1}]",
```

Comment:
```suggestion
        "function \"{0}\" uses the following aspects without declaring them: [{1}]",
```

Reply:
Changed the format to the version from Design Doc.